lib/trino-parquet/src/main/java/io/trino/parquet/BloomFilterStore.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.parquet.hadoop.metadata.ColumnPath;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Objects.requireNonNull;
+
+public class BloomFilterStore
+{
+    // since bloomfilter header is relatively small(18 bytes when testing) we can read in a larger buffer(1024 bytes in this case)
+    // and get actual bytes used when deserializing header in order to calculate the correct offset for bloomfilter data.
+    private static final int MAX_HEADER_LENGTH = 1024;
+
+    private Map<ColumnPath, BloomFilter> bloomFilterStore;
+    private final ParquetDataSource dataSource;
+    private final List<BloomFilterHeaderMetadata> bloomFilterHeaderReferences;
+
+    public BloomFilterStore(ParquetDataSource dataSource, BlockMetaData block, Set<ColumnPath> columnsFiltered)
+    {
+        this.dataSource = requireNonNull(dataSource, "dataSource is null");
+        requireNonNull(block, "block is null");
+        requireNonNull(columnsFiltered, "columnsFiltered is null");
+
+        ImmutableList.Builder<BloomFilterHeaderMetadata> bloomFilterOffsetMapBuilder = ImmutableList.builder();
+        for (ColumnChunkMetaData column : block.getColumns()) {
+            ColumnPath path = column.getPath();
+
+            if (bloomFilterEnabled(column) && columnsFiltered.contains(path)) {
+                bloomFilterOffsetMapBuilder.add(new BloomFilterHeaderMetadata(path, column.getBloomFilterOffset()));
+            }
+        }
+        this.bloomFilterHeaderReferences = bloomFilterOffsetMapBuilder.build();
+    }
+
+    public Optional<BloomFilter> readBloomFilter(ColumnPath columnPath)
+    {
+        if (bloomFilterStore == null) {
+            bloomFilterStore = loadBloomFilters(dataSource, bloomFilterHeaderReferences);
+        }
+
+        return Optional.ofNullable(bloomFilterStore.getOrDefault(columnPath, null));
+    }
+
+    public static boolean bloomFilterEnabled(ColumnChunkMetaData columnMetaData)
+    {
+        return columnMetaData.getBloomFilterOffset() > 0;
+    }
+
+    private static boolean bloomFilterSupported(BloomFilterHeader bloomFilterHeader)
+    {
+        int numBytes = bloomFilterHeader.getNumBytes();
+        if (numBytes <= 0 || numBytes > BlockSplitBloomFilter.UPPER_BOUND_BYTES) {
+            return false;
+        }
+        return bloomFilterHeader.getHash().isSetXXHASH() && bloomFilterHeader.getAlgorithm().isSetBLOCK() && bloomFilterHeader.getCompression().isSetUNCOMPRESSED();
+    }
+
+    private static Map<ColumnPath, BloomFilter> loadBloomFilters(ParquetDataSource dataSource, List<BloomFilterHeaderMetadata> bloomFilterHeaderMetadata)
+    {
+        ListMultimap<ColumnPath, DiskRange> bloomFilterHeaderRanges = ArrayListMultimap.create(bloomFilterHeaderMetadata.size(), 1);
```

Comment:
This is still not ideal because it's likely that we will read bloom filters during the header read step itself due to merging of small and nearby reads inside `dataSource.planRead` but we will not be able to utilise that and make subsequent reads for bloom filter anyway.
Do you know what's the typical size of bloom filter for a row group of around 128MB ?
I see that in parquet-mr the default max size of bloom filters is 1MB and all the bloom filters for a file are written together at the end of the file between the page indexes and the footer (`ParquetFileWriter#end` in parquet-mr).
I'm thinking that for a row group, we can make a sorted array of the bloom filter offsets + the footer start, then we can use that to estimate the length of bloom filter by using the fact that bloom filter length won't exceed the start of the next bloom filter offset. That estimate can be bounded by some threshold like 2MB which is typical size of bloom filter for a column. If after reading bloom filter header from this estimated read it turns out that the estimate was not big enough, we can make another read with the actual size from the header.

Reply:
The size of the bloomfilter depends on two properties: parquet.bloom.filter.expected.ndv and parquet.bloom.filter.fpp. I will provide more details later. Fpp means: The maximum expected false positive probability for the bloom filter in a column.  and NDV means:  The expected number of distinct values in a column, it is used to compute the optimal size of the bloom filter.

The third proposed approach has its heuristics: 1. ordering of the bloomfilter statistics, since in the documentation (https://github.com/apache/parquet-format/blob/master/BloomFilter.md) it didn't specify columnchunk's bloomfilter should be written in the same ordering of the footer. It is possible that it is not sorted 2. the size of per column chunk no larger than 2MB. Due to this, we need to implement retry mechanism when the heuristic2 not holds, and it increased code maintainability. 

Due to reason mentioned above, I will implement: https://github.com/trinodb/trino/pull/14202#issuecomment-1259330709, which only use one round of read.

Reply:
This is an experiment of bloom filter size: https://gist.github.com/leetcode-1533/6b73515f4592381bb795fa0c0b13ee26

The absolute minimum bytes size per bloomfilter is 32 bytes(it is a size of one block, which is composed of 8 words, each words is composed by 1 byte)

The absolute maximum bytes size per bloomfilter is 256MB(268435456 bytes, it is coded as max of 128MB, however, due to rounding to power of 2, can go as high as 256MB). Parquet-mr allows configuring maxBytesSize, however, it cannot go beyond 258MB(it is checked against the UPPER_BOUND_BYTES(128MB) and can go as high as 256 MB due to rounding) 

The sizing is not determined by the sizing of the row group, however, it is determined by number of unique values in the row groups, as well as expected false positive rate. The number of bits is derived from the formula: `double m = -8 * n / Math.log(1 - Math.pow(p, 1.0 / 8))` where n means number of unique values, p means fpp rate. 

When setting fpp as 0.01, this is the chart mapping unique value counts to bloomfilter size in bytes:
|uniqueValues|bytesSize |
|------------|----------|
|1           | 32       |
|100       | 256      |
|1000      | 2048     |
|10000     | 16384    |
|100000    | 131072(128KB)   |
|1000000   | 2097152(2MB) |
|10000000  | 16777216(16MB) |
|100000000 | 134217728(128MB)|
|1000000000| 268435456(256MB)|
|2000000000| 268435456(256MB)|

Reply:
I also debugged how thrift deserialize the parquet bloom filter header, found out it is constant at 18 bytes for the current version, composing of 

schema field: 1 bytes
number of bytes: tag 1 bytes, integer 4 bytes
algorithm: 4 bytes(struct)
hash strategy: 4 bytes(struct)
encryption: 4 bytes(struct)

Reply:
So it looks like bloom filter size can be far bigger than the size of the extra reads that might be planned by `planRead` (merge distance is 1MB by default).
I think we can go with 2 stages of reads that we have right now with reading headers first and then the actual data.
We can use `BlockSplitBloomFilter.HEADER_SIZE * 4` as rough estimate for size of bloom filter header instead of hard-coding 1024.

Reply:
thanks, actively working on publishing the revision, should be done within today or tomorrow.